File: baseclient/java/src/main/java/com/aliyun/oss/baseclient/utils/CRC64.java
package com.aliyun.oss.baseclient.utils;


import java.math.BigDecimal;

/**
 * CRC-64 implementation with ability to combine checksums calculated over
 * different blocks of data. Standard ECMA-182,
 * http://www.ecma-international.org/publications/standards/Ecma-182.htm
 */
public class CRC64 {
    // ECMA-182
    private final static long POLY = (long) 0xc96c5795d7870f42L;

    /* CRC64 calculation table. */
    private final static long[] table;

    /* Current CRC value. */
    private long value;

    static {
        table = new long[256];
        for (int n = 0; n < 256; n++) {
            long crc = n;
            for (int k = 0; k < 8; k++) {
                if ((crc & 1) == 1) {
                    crc = (crc >>> 1) ^ POLY;
                } else {
                    crc = (crc >>> 1);
                }
            }
            table[n] = crc;
        }
    }

    public CRC64() {
        this.value = 0;
    }


    public String getValue() {
        long first = (this.value >>> 1) / 5;
        long second = this.value - first * 10;
        return String.valueOf(first) + second;
    }



    /**
     * Update CRC64 with new byte block.
     **/
    public void update(byte[] b, int len) {
        for (int i = 0; i < len; i++) {
            update(b[i]);
        }
    }


    /**
     * Update CRC64 with new byte.
     **/
    public void update(byte b) {
        this.value = ~this.value;
        this.value = table[((int) (this.value ^ b)) & 0xff] ^ (this.value >>> 8);
        this.value = ~this.value;
    }


    public void reset() {
        this.value = 0;
    }

}

File: util/swift/Sources/AlibabaCloudOSSUtil/CRC64.swift
import Foundation

open class CRC64 {
    // ECMA-182
    private static let POLY: UInt64 = 0xc96c5795d7870f42
    private static var table = [UInt64](repeating: 0, count: 256)
    private var value: UInt64 = 0
    
    static func initializeTable() {
        for n in 0..<256 {
            var crc = UInt64(n)
            for _ in 0..<8 {
                if crc & 1 == 1 {
                    crc = (crc >> 1) ^ POLY
                } else {
                    crc = crc >> 1
                }
            }
            table[n] = crc
        }
    }
    
    init() {
        self.value = 0
        CRC64.initializeTable()
    }
    
    func getValue() -> String {
        let first = (self.value >> 1) / 5
        let second = self.value - first * 10
        return "\(first)\(second)"
    }
    
    func update(_ b: [UInt8], len: Int) {
        for i in 0..<len {
            update(b[i])
        }
    }
    
    func update(_ b: UInt8) {
        self.value = ~self.value
        self.value = CRC64.table[Int((self.value ^ UInt64(b)) & 0xff)] ^ (self.value >> 8)
        self.value = ~self.value
    }
}
